text/0088-fix-memory-limitiations-in-session-replays-access-pattern.md

```diff
@@ -0,0 +1,324 @@
+    - This adds complexity to the consumer but not an unmanageable amount.
+    - The replay-events consumer is a Snuba consumer so we may encounter some minor procedural challenges that a consumer wholly owned by the replays team would not.
+    - Replays are always "live" for one hour. They are not eagerly closed.
+    - Because this uses I/O it will block our snuba consumer from running at peak efficiency. Should the message be forwarded to another topic after order has been guaranteed?
+- Bruno Garcia asks: The complexity of integrating a new message queue (RabbitMQ in the above answer) into our ingest pipeline seems high. What sort of scaling issues arise, how do we handle outages of each service component, can a cron job (using a sliding window query) simplify?
+  - Scale: The message throughput for RabbitMQ will be roughly 1/10th of what we process on our Kafka consumer (we only process one message per replay_id). The messages will be small and the Celery task's only function will be to forward the message back to the Kafka consumer. The load recieved by RabbitMQ and the task process should be very managable. Additionally, there are other bottlenecks in the system which will take precedence.
+  - Outages: If Relay is down new messages are rejected and the queues are drained. If the Kafka consumer is down then the messages backlog until the consumer resumes processing. If RabbitMQ is down the Kafka consumer must pause.
+  - Simplify with cron job: If the cron job has an outage it can be resumed from a save point. However, if the Kafka consumer is down then we need to pause the cron so that it does not write incomplete aggregation rows. Both use cases require a piece of infrastructure to be paused in the event of an outage.
+    - If our Kafka consumer encounters a RabbitMQ failure it can pause processing, publish the message to a DLQ, or publish the message back to itself to retry automatically at a later time.
+
+### 11. Configure ClickHouse to Optimize Aggregations
+
+**Proposal**
+
+Use `optimize_aggregation_in_order` to limit the number of rows we need to aggregate. In testing this eliminates OOM errors. Queries complete somewhat speedily.
+
+**Drawbacks**
+
+**Questions**
+
+### 12. Aggregate Replay Event Metadata in the SDK and Set Final Flag
+
+**Proposal**
+
+The SDK can maintain a buffer of the replay metadata. Once the replay has finished the SDK will flush that buffered data on the final segment. The segment marked as "final" can then be fetched without aggregation.
+
+**Drawbacks**
+
+- Requies SDK upgrade.
+- API will need to fallback to aggregating behavior if no final segments can be found (or otherwise detect old SDK usage prior to querying).
+
+**Questions**
+
+- Is it possible for the SDK to know when its done and mark the request object with a final attribute?
```

Comment:
@billyvg Can we assess this?  Is it possible for the SDK to _know_ when the replay has finished?  If so do we have any idea how reliable it is?